---
title: "freqSequences"
output: html_document
---
```{r}
library(stats)
library(shiny)
library(shinyFiles)
library(ShortRead)
library(shinyjs)
library(dplyr)
library(ggplot2)
library(Biostrings)
library(gridExtra)
library(grid)

arquivo_fastq <- "C:/Users/vitor/Documents/ufpr/projeto-bioinformatica/fastq_gerados/output2.fastq"
arquivo_fastq <- "C:/Users/vitor/Documents/ufpr/projeto-bioinformatica/teste_funcoes/adapter_contamination/mock.fastq"
arquivo_fastq <- "C:/Users/vitor/Documents/ufpr/projeto-bioinformatica/teste_funcoes/adapter_contamination/mock2.fastq"

freqSequences <- function(qa, read, n=20)
{
    cnt <- qa[["readCounts"]]
    df <- qa[["frequentSequences"]]
    df1 <- df[df$type==read,]
    df1[["ppn"]] <- df1[["count"]] / cnt[df1[["lane"]], read]
    df <- head(df1[order(df1$count, decreasing=TRUE),
                   c("sequence", "count", "lane")], n)
    rownames(df) <- NULL
    df
}

qaSummary <- qa(arquivo_fastq, type = "fastq")

df <- qaSummary[["sequenceDistribution"]]
freqsequences <-  freqSequences(qaSummary, read="read", n=20)
freqsequences$lane <- NULL

freqsequences$possible_hit <- sapply(freqsequences$sequence, function(seq) {
  matches <- vcountPattern(DNAString(seq), adapter_fasta, fixed=FALSE)
  if (any(matches > 0)) {
    names(adapter_fasta)[which.max(matches)]  # return the adapter name with most hits
  } else {
    "no hit"
  }
})



# Set PNG output
png("frequent_sequences_table.png", width = 1000, height = 400)

# Draw table
grid.table(freqsequences, rows = NULL) 

# Close the device
dev.off()
```